### PR TITLE
Run Hooks Improvements

### PR DESCRIPTION
**Pull request by terrytangyuan**
- Removed `verbose` and `view_metrics` from `train()` and `evaluate()` so users can optionally attach them as hooks.
- Modified `hook_progress_bar`, `hook_history_saver`, and `hook_view_metrics` so that when if users pass them to `train` or `evaluate` certain necessary metadata can be attached accordingly.
- Exposed `visualize_metrics()` to visualize the metrics produced during training or evaluation.
- Rewrote run hook tutorial to show usages of `hook_progress_bar` and `hook_history_saver` as well as an example of implementing custom run hook `hook_history_saver` using `session_run_hook()`. 

Note that `hook_view_metrics` is broken and I am still looking into it. For now users can use `visualize_metrics` and in the meantime, we can also think about whether real-time visualization is necessary. I believe `visualize_metrics` should be sufficient for now as it doesn't introduces any overhead. It's a little tricky to chain `hook_history_saver` and `hook_view_metrics` together.

### CODE REVIEW

File: R/session_run_hooks_custom.R [hunk 1]
```diff
@@ -1,134 +1,226 @@
...
+
+#' Visualize the training or evaluation metrics
+#' 
+#' This function allows users to visualize the metrics produced in training or evaluation phases.
+#' 
+#' @param mode_key The mode when the metrics were collected that you want to visualize, e.g. "train"
+#' 
+#' @return The directory used to save the metrics metadata and generated html file.
+#' @export
+visualize_metrics <- function(mode_key = c("train", "eval")) {
```
Comment (jjallaire) [inline R/session_run_hooks_custom.R]: I think we should get rid of this function as it's only a temporary hack until we figure out why `hook_view_metrics doesn't work). We can just remove `hook_view_metrics()` entirely until we sort all of that out.

Comment (terrytangyuan) [reply R/session_run_hooks_custom.R]: Do you mean removing visualize_metrics() as well?

Comment (jjallaire) [reply R/session_run_hooks_custom.R]: Yes, I don't think we want to support that longer term so better we leave it out for now.

Comment (terrytangyuan) [reply R/session_run_hooks_custom.R]: Removed and added a function `get_saved_history` in https://github.com/rstudio/tfestimators/pull/111/commits/1be2d5593b1f3043e3b7b5e855dc1c5d46d73ad2.

**Comment by terrytangyuan** [comment]
@jjallaire We had history as part of the model object but I forgot to add it back! I just pushed some changes addressing your comments.

**Comment by jjallaire** [comment]
Okay, great!

What do you think of returning the history from the `train` method rather than returning the model object? (that would allow for an interface more similar to Keras). I noticed that the Python interface doesn't return anything from train (or perhaps it does and we don't currently capture it). It seems unlikely that users will chain the `train` method directly with another method after it.

**Comment by terrytangyuan** [comment]
That's a good suggestion. I just changed that to be consistent with Keras. Travis build seems to be failing under nightly builds though.

**Comment by terrytangyuan** [comment]
I opened a separate issue to track that since it's irrelevant to this PR: https://github.com/rstudio/tfestimators/issues/113

File: docs/reference/train.tf_estimator.html [hunk 2]
```diff
@@ -171,6 +171,11 @@ <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a> Ar
     </tr>
     </table>
     
+    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>
+
+    <p>A data.frame of the training loss history. Note that this will be
+a empty data.frame if <code><a href='hook_history_saver.html'>hook_history_saver()</a></code> is not used as one of the hooks provided.</p>
```
Comment (jjallaire) [inline docs/reference/train.tf_estimator.html]: Should we attach the `hook_history_saver()` by default? (seems like just accumulating the metrics shouldn't have any performance problems)

Comment (terrytangyuan) [reply docs/reference/train.tf_estimator.html]: We could but how frequent should we save? Right now we have `every_n_step` but in the future it's possible to add things like `every_n_second`. What should the default be?

Comment (jjallaire) [reply docs/reference/train.tf_estimator.html]: If all the history does is update some vectors it seems like it's not a big deal to save it frequently (the worst that can happen is too many points to plot). Maybe every 10 steps?

Comment (jjallaire) [reply docs/reference/train.tf_estimator.html]: Is the progress bar hook also automatically applied? (it is for Keras).

Comment (terrytangyuan) [reply docs/reference/train.tf_estimator.html]: Got it. I just addressed those in https://github.com/rstudio/tfestimators/pull/111/commits/42ac84633d375f11a022301c9311b27cc2c24721!

**Comment by jjallaire** [comment]
Okay, great! 

Seems like this PR might be ready to merge (modulo the failing tests). Do you agree?

**Comment by terrytangyuan** [comment]
Yea I'll merge it. The failing test is irrelevant to this PR (same failure in other PR) and I'll look into it in a bit.

**Comment by terrytangyuan** [comment]
Thanks for your careful and helpful reviews!

**Pull request merged**